File: lib/features/outfit_analysis/domain/helpers/outfit_fashion_taxonomy.dart
import 'package:flutter/material.dart';

import '../entities/outfit_segment_map.dart';
import '../services/outfit_segmentation_service.dart';

/// Fashion object classification for Vision-first segmentation.
abstract final class OutfitFashionTaxonomy {
  OutfitFashionTaxonomy._();

  static const minVisionScore = 0.82;
  static const minDisplayConfidence = 0.82;

  static const clothingKeywords = [
    'dress',
    'shirt',
    't-shirt',
    'blouse',
    'top',
    'blazer',
    'jacket',
    'coat',
    'cape',
    'corset',
    'suit',
    'pants',
    'trouser',
    'jean',
    'skirt',
    'sweater',
    'hoodie',
    'outerwear',
    'cardigan',
    'gown',
    'formal',
  ];

  static const footwearKeywords = ['shoe', 'heel', 'boot', 'sneaker', 'footwear'];

  static const bagKeywords = ['bag', 'handbag', 'purse', 'backpack', 'clutch'];

  static const accessoryKeywords = [
    'watch',
    'sunglass',
    'glasses',
    'belt',
    'scarf',
    'hat',
    'jewelry',
    'necklace',
    'earring',
    'bracelet',
  ];

  static const hiddenOverlayZones = {
    OutfitSegmentZone.head,
    OutfitSegmentZone.waist,
  };

  static bool isFashionObject(String name) {
    final lower = name.toLowerCase();
    if (lower.contains('person') || lower.contains('human') || lower.contains('face')) {
      return false;
    }
    return isClothing(name) ||
        isFootwear(name) ||
        isBag(name) ||
        isAccessory(name);
  }

  static bool isClothing(String name) {
    final lower = name.toLowerCase();
    return clothingKeywords.any(lower.contains);
  }

  static bool isFootwear(String name) {
    final lower = name.toLowerCase();
    return footwearKeywords.any(lower.contains);
  }

  static bool isBag(String name) {
    final lower = name.toLowerCase();
    return bagKeywords.any(lower.contains) || lower.contains('حقيب');
  }

  static bool isAccessory(String name) {
    final lower = name.toLowerCase();
    return accessoryKeywords.any(lower.contains) ||
        lower.contains('إكسسوار') ||
        lower.contains('ساع') ||
        lower.contains('نظار') ||
        lower.contains('وشاح');
  }

  static OutfitSegmentZone zoneForObject(String name, Rect box) {
    if (isFootwear(name)) return OutfitSegmentZone.feet;
    if (isBag(name) || isAccessory(name)) return OutfitSegmentZone.accessories;
    if (isClothing(name)) {
      final centerY = box.center.dy;
      if (centerY >= 0.52) return OutfitSegmentZone.lowerBody;
      if (centerY >= 0.38 && centerY < 0.52) return OutfitSegmentZone.waist;
      return OutfitSegmentZone.upperBody;
    }
    final centerY = box.center.dy;
    if (centerY >= 0.78) return OutfitSegmentZone.feet;
    if (centerY >= 0.48) return OutfitSegmentZone.lowerBody;
    if (centerY >= 0.38) return OutfitSegmentZone.waist;
    return OutfitSegmentZone.upperBody;
  }

  static List<VisionLocalizedObject> filterFashionObjects(
    List<VisionLocalizedObject> objects,
  ) {
    final filtered = objects
        .where((o) => isFashionObject(o.name) && o.score >= minVisionScore)
        .toList()
      ..sort((a, b) => b.score.compareTo(a.score));
    return filtered;
  }

  static bool isGenericPlaceholderLabel(OutfitSegmentRegion region) {
    final ar = region.labelAr;
    final en = region.labelEn.toLowerCase();
    return ar == 'الجزء العلوي' ||
        ar == 'الجزء السفلي' ||
        ar == 'الحذاء' ||
        ar == 'حذاء' ||
        ar == 'غير مؤكد' ||
        ar == 'إكسسوار' ||
        en == 'upper body' ||
        en == 'lower body' ||
        en == 'shoes' ||
        en == 'accessory';
  }

  static bool shouldShowRegion(OutfitSegmentRegion region) {
    if (hiddenOverlayZones.contains(region.zone)) return false;
    if (isGenericPlaceholderLabel(region)) return false;
    if (region.confidence < minDisplayConfidence) return false;

    if (region.zone == OutfitSegmentZone.accessories) {
      final lower = region.labelEn.toLowerCase();
      return isBag(lower) ||
          isAccessory(lower) ||
          region.labelAr.contains('حقيب') ||
          region.labelAr.contains('عقد') ||
          region.labelAr.contains('قلادة') ||
          region.labelAr.contains('ساعة') ||
          region.labelAr.contains('نظارة');
    }
    if (region.zone == OutfitSegmentZone.feet) {
      return isFootwear(region.labelEn) || region.labelAr.contains('حذ');
    }
    return region.labelAr.isNotEmpty;
  }

  static IconData iconForPiece(String labelAr) {
    final lower = labelAr.toLowerCase();
    if (lower.contains('حذ') || lower.contains('heel') || lower.contains('shoe')) {
      return Icons.directions_walk_rounded;
    }
    if (lower.contains('حقيب') || lower.contains('bag')) {
      return Icons.shopping_bag_outlined;
    }
    if (lower.contains('بنط') || lower.contains('جين') || lower.contains('jean')) {
      return Icons.straighten_rounded;
    }
    if (lower.contains('فستان') || lower.contains('dress')) {
      return Icons.woman_rounded;
    }
    if (lower.contains('تنورة') || lower.contains('skirt')) {
      return Icons.line_style_rounded;
    }
    if (lower.contains('جاك') || lower.contains('بلوز') || lower.contains('ستر') ||
        lower.contains('blazer') || lower.contains('jacket')) {
      return Icons.checkroom_outlined;
    }
    if (lower.contains('تيش') || lower.contains('shirt') || lower.contains('قمي')) {
      return Icons.checkroom_outlined;
    }
    if (lower.contains('ساع')) return Icons.watch_outlined;
    if (lower.contains('نظار')) return Icons.visibility_outlined;
    if (lower.contains('وشاح')) return Icons.air_rounded;
    return Icons.checkroom_outlined;
  }

  static IconData iconForAccessory(String label) {
    final lower = label.toLowerCase();
    if (lower.contains('ساع')) return Icons.watch_outlined;
    if (lower.contains('نظار')) return Icons.visibility_outlined;
    if (lower.contains('حقيب')) return Icons.shopping_bag_outlined;
    if (lower.contains('وشاح')) return Icons.air_rounded;
    if (lower.contains('حزام')) return Icons.linear_scale_rounded;
    if (lower.contains('أقر') || lower.contains('قلادة')) {
      return Icons.diamond_outlined;
    }
    return Icons.auto_awesome_outlined;
  }

  static String verdictForScore(int score) {
    if (score >= 90) return 'نتيجة الإطلالة استثنائية';
    if (score >= 85) return 'نتيجة الإطلالة ممتاز';
    if (score >= 75) return 'نتيجة الإطلالة جيدة جداً';
    if (score >= 65) return 'نتيجة الإطلالة جيدة';
    if (score >= 50) return 'نتيجة الإطلالة مقبولة';
    return 'تحتاجين تحسينات';
  }

  static List<OutfitSegmentRegion> visibleRegions(List<OutfitSegmentRegion> regions) {
    return regions.where(shouldShowRegion).toList();
  }
}

File: mira-api/src/ai/segmentation/outfit-fashion-taxonomy.ts
import { OutfitSegmentZone } from '../contracts/outfit-segment-map.interface';

const CLOTHING = [
  'dress',
  'shirt',
  't-shirt',
  'blouse',
  'top',
  'blazer',
  'jacket',
  'coat',
  'suit',
  'pants',
  'trouser',
  'jean',
  'skirt',
  'sweater',
  'hoodie',
  'outerwear',
];

const FOOTWEAR = ['shoe', 'heel', 'boot', 'sneaker', 'footwear'];
const BAG = ['bag', 'handbag', 'purse', 'backpack', 'clutch'];
const ACCESSORY = [
  'watch',
  'sunglass',
  'glasses',
  'belt',
  'scarf',
  'hat',
  'jewelry',
  'necklace',
  'earring',
  'bracelet',
];

export const MIN_VISION_SCORE = 0.42;
export const MIN_DISPLAY_CONFIDENCE = 0.4;

export function isFashionObject(name: string): boolean {
  const lower = name.toLowerCase();
  if (lower.includes('person') || lower.includes('human') || lower.includes('face')) {
    return false;
  }
  return (
    isClothing(lower) ||
    isFootwear(lower) ||
    isBag(lower) ||
    isAccessory(lower)
  );
}

export function isClothing(lower: string): boolean {
  return CLOTHING.some((k) => lower.includes(k));
}

export function isFootwear(lower: string): boolean {
  return FOOTWEAR.some((k) => lower.includes(k));
}

export function isBag(lower: string): boolean {
  return BAG.some((k) => lower.includes(k)) || lower.includes('حقيب');
}

export function isAccessory(lower: string): boolean {
  return (
    ACCESSORY.some((k) => lower.includes(k)) ||
    lower.includes('إكسسوار') ||
    lower.includes('ساع') ||
    lower.includes('نظار')
  );
}

export function zoneForObject(name: string, centerY: number): OutfitSegmentZone {
  const lower = name.toLowerCase();
  if (lower.includes('dress') || lower.includes('gown')) return 'upperBody';
  if (isFootwear(lower)) return 'feet';
  if (isBag(lower) || isAccessory(lower)) return 'accessories';
  if (isClothing(lower)) {
    if (centerY >= 0.52) return 'lowerBody';
    if (centerY >= 0.38) return 'waist';
    return 'upperBody';
  }
  if (centerY >= 0.78) return 'feet';
  if (centerY >= 0.48) return 'lowerBody';
  if (centerY >= 0.38) return 'waist';
  return 'upperBody';
}

export function labelToArabic(label: string): string {
  const lower = label.toLowerCase();
  if (lower.includes('dress')) return 'فستان';
  if (lower.includes('abaya')) return 'عباءة';
  if (lower.includes('suit')) return 'بدلة';
  if (lower.includes('skirt')) return 'تنورة';
  if (lower.includes('t-shirt') || lower.includes('tee')) return 'تيشيرت';
  if (lower.includes('blouse')) return 'بلوزة';
  if (lower.includes('blazer')) return 'بلوزر';
  if (lower.includes('shirt')) return 'قميص';
  if (lower.includes('pants') || lower.includes('trouser') || lower.includes('jean')) {
    return 'بنطلون';
  }
  if (lower.includes('jacket') || lower.includes('coat')) return 'جاكيت';
  if (lower.includes('bag') || lower.includes('handbag')) return 'حقيبة';
  if (lower.includes('shoe') || lower.includes('heel') || lower.includes('boot')) {
    return 'حذاء';
  }
  if (lower.includes('watch')) return 'ساعة';
  if (lower.includes('sunglass') || lower.includes('glasses')) return 'نظارة';
  return label;
}

export function shouldShowRegion(region: {
  zone: OutfitSegmentZone;
  labelEn: string;
  labelAr: string;
  confidence: number;
}): boolean {
  if (region.zone === 'head' || region.zone === 'waist') return false;
  if (region.zone === 'accessories') {
    const lower = region.labelEn.toLowerCase();
    const ok =
      isBag(lower) ||
      isAccessory(lower) ||
      region.labelAr.includes('حقيب') ||
      region.labelAr.includes('ساع') ||
      region.labelAr.includes('نظارة');
    return ok && region.confidence >= MIN_DISPLAY_CONFIDENCE;
  }
  if (region.zone === 'feet') {
    return region.confidence >= MIN_DISPLAY_CONFIDENCE || isFootwear(region.labelEn);
  }
  return region.confidence >= 0.35;
}
